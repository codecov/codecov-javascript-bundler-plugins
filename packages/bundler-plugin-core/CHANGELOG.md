# @codecov/bundler-plugin-core

## 0.0.1-beta.0

### Patch Changes

- f5cae1c: Move packages into beta

## 0.0.1-beta.0

### Patch Changes

- becd728: Add duration to webpack bundler plugin directly
- 543a526: Update dependencies before moving package out of alpha
- 4f1183e: Remove references to Webpack 4, as we currently on support Node 18+.
- c1fdbd6: Fix retry when fetching by throwing custom error if response is not okay
- d28a2e4: Break from retry loop when the response is ok
- 7c55993: Rename uploaderOverrides to uploadOverrides, remove url from uploadOverrides, and removal of repoToken/globalUploadToken in favour of just uploadToken.
- 36ed299: Alpha release of Codecov bundler plugins
- 562ac0c: Adjust peer dep versions to use x-range instead of caret
- 48b6e90: Move bundler-core test deps to dev deps
- c755a5c: Add missing outputPath to output stats file

## 0.0.1-alpha.3

### Patch Changes

- 543a526: Update dependencies before moving package out of alpha
- c1fdbd6: Fix retry when fetching by throwing custom error if response is not okay
- d28a2e4: Break from retry loop when the response is ok
- 562ac0c: Adjust peer dep versions to use x-range instead of caret
- c755a5c: Add missing outputPath to output stats file

## 0.0.1-alpha.2

### Patch Changes

- becd728: Add duration to webpack bundler plugin directly
- 7c55993: Rename uploaderOverrides to uploadOverrides, remove url from uploadOverrides, and removal of repoToken/globalUploadToken in favour of just uploadToken.

## 0.0.1-alpha.1

### Patch Changes

- 4f1183e: Remove references to Webpack 4, as we currently on support Node 18+.
- 48b6e90: Move bundler-core test deps to dev deps

## 0.0.1-alpha.0

### Patch Changes

- 36ed299: Alpha release of Codecov bundler plugins
