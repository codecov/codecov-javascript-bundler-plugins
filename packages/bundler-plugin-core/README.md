<p align="center">
  <a href="https://about.codecov.io" target="_blank">
    <img src="https://about.codecov.io/wp-content/themes/codecov/assets/brand/sentry-cobranding/logos/codecov-by-sentry-logo.svg" alt="Codecov by Sentry logo" width="280" height="84">
  </a>
</p>

# Codecov Bundler Plugin Core

> [!WARNING]  
> These plugins are currently in beta and are subject to change.

> The bundler-plugin-core is not to be used individually and is meant to be used in conjunction with the bundler specific plugins.

Core package containing the bundler-agnostic logic for the Codecov Bundler Plugins.

Checkout the individual packages for more information and examples:

- [Rollup](https://npmjs.org/package/@codecov/rollup-plugin)
- [Vite](https://npmjs.org/package/@codecov/vite-plugin)
- [Webpack](https://npmjs.org/package/@codecov/webpack-plugin)

## Features

The Codecov Bundler Core package contains the following functionality:

- Bundle Analysis Stats Upload

## More information

- [Bundler Config Docs](https://codecov.github.io/codecov-javascript-bundler-plugins/modules/_codecov_bundler_plugin_core.html)
- [Codecov Documentation](https://docs.codecov.com/docs)
- [Codecov Feedback](https://github.com/codecov/feedback/discussions)
- [Sentry Discord](https://discord.gg/Ww9hbqr)
