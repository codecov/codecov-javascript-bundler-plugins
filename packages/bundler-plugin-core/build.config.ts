import { defineBuildConfig } from "unbuild";
import { codecovRollupPlugin } from "codecovProdRollupPlugin";
import packageJson from "./package.json";

export default defineBuildConfig({
  entries: ["./src/index"],
  outDir: "dist",
  declaration: "compatible",
  sourcemap: true,
  rollup: {
    dts: {
      compilerOptions: {
        removeComments: false,
      },
    },
    emitCJS: true,
    esbuild: {
      minify: true,
    },
  },
  hooks: {
    "rollup:options": (_ctx, opts) => {
      if (process.env.PLUGIN_CODECOV_TOKEN && Array.isArray(opts.plugins)) {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call
        opts.plugins = [
          ...opts.plugins,
          codecovRollupPlugin({
            enableBundleAnalysis:
              typeof process.env.PLUGIN_CODECOV_TOKEN === "string",
            bundleName: packageJson.name,
            uploadToken: process.env.PLUGIN_CODECOV_TOKEN,
            apiUrl: process.env.PLUGIN_CODECOV_API_URL,
          }),
        ];
      }
    },
  },
});
