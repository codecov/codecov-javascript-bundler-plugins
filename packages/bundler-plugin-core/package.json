{
  "name": "@codecov/bundler-plugin-core",
  "version": "0.0.1-beta.3",
  "description": "Official Codecov Bundler Plugin Core",
  "author": "Codecov",
  "license": "MIT",
  "type": "module",
  "exports": {
    ".": {
      "import": {
        "types": "./dist/index.d.mts",
        "default": "./dist/index.mjs"
      },
      "require": {
        "types": "./dist/index.d.ts",
        "default": "./dist/index.cjs"
      }
    }
  },
  "main": "./dist/index.cjs",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "type-check": "tsc --noEmit",
    "build": "unbuild",
    "dev": "unbuild --stub && node --watch-path=src dist/index.mjs",
    "clean": "rm -rf dist node_modules",
    "lint": "eslint . --ext .ts,.tsx",
    "lint:fix": "pnpm lint --fix",
    "format": "prettier '**/*.{cjs,mjs,ts,tsx,md,json}' --ignore-path ../.gitignore --ignore-unknown --no-error-on-unmatched-pattern --write",
    "format:check": "prettier '**/*.{cjs,mjs,ts,tsx,md,json}' --ignore-path ../.gitignore --ignore-unknown --no-error-on-unmatched-pattern --check",
    "test:unit": "jest",
    "test:unit:ci": "jest --coverage"
  },
  "dependencies": {
    "chalk": "4.1.2",
    "semver": "^7.5.4",
    "unplugin": "^1.6.0",
    "zod": "^3.22.4"
  },
  "devDependencies": {
    "@swc/core": "^1.3.107",
    "@swc/jest": "^0.2.33",
    "@types/jest": "^29.5.11",
    "@types/node": "^20.11.15",
    "@types/semver": "^7.5.6",
    "jest": "^29.7.0",
    "msw": "^2.1.5",
    "codecovProdRollupPlugin": "npm:@codecov/rollup-plugin@0.0.1-beta.2",
    "testdouble": "^3.20.1",
    "testdouble-jest": "^2.0.0",
    "ts-node": "^10.9.2",
    "typescript": "^5.3.3",
    "unbuild": "^2.0.0"
  },
  "volta": {
    "extends": "../../package.json"
  },
  "engines": {
    "node": ">=18.0.0"
  }
}
