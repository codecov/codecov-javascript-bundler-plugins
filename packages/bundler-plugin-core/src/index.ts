import {
  type Asset,
  type Chunk,
  type Module,
  type Options,
  type ProviderUtilInputs,
  type UploadOverrides,
  type BundleAnalysisUploadPlugin,
} from "./types.ts";
import { checkNodeVersion } from "./utils/checkNodeVersion.ts";
import { red } from "./utils/logging.ts";
import { normalizeOptions } from "./utils/normalizeOptions.ts";
import { normalizePath } from "./utils/normalizePath.ts";
import { Output } from "./utils/Output.ts";

export type {
  Asset,
  Chunk,
  Module,
  Options,
  ProviderUtilInputs,
  UploadOverrides,
  BundleAnalysisUploadPlugin,
};

export { checkNodeVersion, normalizeOptions, normalizePath, red, Output };
