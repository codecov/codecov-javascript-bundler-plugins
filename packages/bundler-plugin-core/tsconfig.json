{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "baseUrl": "./",
    "paths": {
      "@/*": ["./src/*"],
      "@test-utils/*": ["./test-utils/*"]
    },
    "checkJs": true
  },
  "include": [
    "src",
    "test-utils",
    "build.config.ts",
    "../../reset.d.ts",
    "jest.config.ts"
  ]
}
