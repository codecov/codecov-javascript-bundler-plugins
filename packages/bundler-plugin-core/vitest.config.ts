import { defineConfig } from "vitest/config";
import { config } from "../../vitest.shared";

export default defineConfig({
  ...config,
});
