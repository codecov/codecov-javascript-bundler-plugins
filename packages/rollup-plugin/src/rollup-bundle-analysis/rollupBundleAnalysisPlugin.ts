import path from "node:path";
import {
  type Asset,
  type Chunk,
  type Module,
  type BundleAnalysisUploadPlugin,
  red,
} from "@codecov/bundler-plugin-core";

const PLUGIN_NAME = "codecov-rollup-bundle-analysis-plugin";

export const rollupBundleAnalysisPlugin: BundleAnalysisUploadPlugin = ({
  output,
  userOptions,
}) => ({
  version: "1",
  name: PLUGIN_NAME,
  pluginVersion: "1.0.0",
  rollup: {
    generateBundle(this, options, bundle) {
      // don't need to do anything if the bundle name is not present or empty
      if (!userOptions.bundleName || userOptions.bundleName === "") {
        red("Bundle name is not present or empty. Skipping upload.");
        return;
      }

      // append bundle output format to bundle name
      output.bundleName = `${userOptions.bundleName}-${options.format}`;

      if (options.name && options.name !== "") {
        output.bundleName = `${userOptions.bundleName}-${options.name}`;
      }

      const customOptions = {
        moduleOriginalSize: false,
        ...options,
      };

      const assets: Asset[] = [];
      const chunks: Chunk[] = [];
      const moduleByFileName = new Map<string, Module>();
      const items = Object.values(bundle);

      const cwd = process.cwd();

      let counter = 0;
      for (const item of items) {
        if (item?.type === "asset") {
          if (typeof item.source === "string") {
            const fileName = item?.fileName ?? "";
            const size = Buffer.from(item.source).byteLength;

            assets.push({
              name: fileName,
              size: size,
            });
          } else {
            const fileName = item?.fileName ?? "";
            const size = item?.source?.byteLength;

            assets.push({
              name: fileName,
              size: size,
            });
          }
        }

        if (item?.type === "chunk") {
          const chunkId = item?.name ?? "";
          const fileName = item?.fileName ?? "";
          const moduleEntries = Object.entries(item?.modules ?? {});
          const size = item?.code?.length;
          const uniqueId = `${counter}-${chunkId}`;

          assets.push({
            name: fileName,
            size: size,
          });

          chunks.push({
            id: chunkId,
            uniqueId: uniqueId,
            entry: item?.isEntry,
            initial: item?.isDynamicEntry,
            files: [item?.fileName],
            names: [item?.name],
          });

          for (const [modulePath, moduleInfo] of moduleEntries) {
            const normalizedModulePath = modulePath.replace("\u0000", "");
            const relativeModulePath = path.relative(cwd, normalizedModulePath);
            const relativeModulePathWithPrefix = relativeModulePath.match(
              /^\.\./,
            )
              ? relativeModulePath
              : `.${path.sep}${relativeModulePath}`;

            // try to grab module already set in map
            const moduleEntry = moduleByFileName.get(
              relativeModulePathWithPrefix,
            );

            // if the modules exists append chunk ids to the grabbed module
            // else create a new module and create a new entry in the map
            if (moduleEntry) {
              moduleEntry.chunks.push(chunkId);
              moduleEntry.chunkUniqueIds.push(uniqueId);
            } else {
              const size = customOptions.moduleOriginalSize
                ? moduleInfo.originalLength
                : moduleInfo.renderedLength;

              const module = {
                name: relativeModulePathWithPrefix,
                size: size,
                chunks: [chunkId],
                chunkUniqueIds: [uniqueId],
              };

              moduleByFileName.set(relativeModulePathWithPrefix, module);
            }
          }
          counter += 1;
        }
      }

      // grab the modules from the map and convert it to an array
      const modules = Array.from(moduleByFileName.values());

      output.bundler = {
        name: "rollup",
        version: this.meta.rollupVersion,
      };
      output.assets = assets;
      output.chunks = chunks;
      output.modules = modules;

      // only output file if running dry run
      if (userOptions?.dryRun) {
        this.emitFile({
          type: "asset",
          fileName: "codecov-bundle-stats.json",
          source: JSON.stringify(output),
        });
      }
    },
  },
});
