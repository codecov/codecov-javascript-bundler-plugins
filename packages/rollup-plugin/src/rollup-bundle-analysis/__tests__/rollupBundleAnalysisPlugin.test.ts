import { Output } from "@codecov/bundler-plugin-core";
import { describe, it, expect } from "vitest";
import { rollupBundleAnalysisPlugin } from "../rollupBundleAnalysisPlugin";

describe("rollupBundleAnalysisPlugin", () => {
  describe("when called", () => {
    it("returns a plugin object", () => {
      const plugin = rollupBundleAnalysisPlugin({
        output: new Output({
          apiUrl: "http://localhost",
          bundleName: "test-bundle",
          debug: false,
          dryRun: true,
          enableBundleAnalysis: true,
          retryCount: 1,
          uploadToken: "test-token",
        }),
      });

      expect(plugin).toMatchSnapshot();
    });
  });
});
