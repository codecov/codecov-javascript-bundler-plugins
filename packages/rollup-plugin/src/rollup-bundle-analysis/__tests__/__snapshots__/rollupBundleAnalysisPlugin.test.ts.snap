// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rollupBundleAnalysisPlugin when called returns a plugin object 1`] = `
{
  "buildEnd": [Function],
  "buildStart": [Function],
  "name": "@codecov/rollup-plugin",
  "pluginVersion": "0.0.1-beta.6",
  "rollup": {
    "generateBundle": [Function],
  },
  "version": "1",
  "writeBundle": [Function],
}
`;
